### DISCUSSION TITLE
Kupo Match Patterns

### DISCUSSION THREAD

**Discussion by mrosberghaus** [geniusyield/smart-order-router#7] [Q&A] [answered]
I am unsure if I am using the correct matching pattern for the smart-order-router. I hope it's correct that kupo is indexing transactions to the `refAddr` and then checks the exact data to look for matching orders. Is the assumption correct or should I index everything with `*`?

I defined in two matches:
- _refAddr_: `--match=addr_test1wpgexmeunzsykesf42d4eqet5yvzeap6trjnflxqtkcf66g0kpnxt`
- _refNftAC_: `--match=fae686ea8f21d567841d703dea4d4221c2af071a6f2b433ff07c0af2.e6a295bb83d06f53fcf91151f54acec0a63fbd6f0d924206d5d012e6da3b72af`

Based on the validationRefs:
```json
{
"validatorRefs":
  { "refAddr": "addr_test1wpgexmeunzsykesf42d4eqet5yvzeap6trjnflxqtkcf66g0kpnxt"
  , "refNftAC": "fae686ea8f21d567841d703dea4d4221c2af071a6f2b433ff07c0af2.e6a295bb83d06f53fcf91151f54acec0a63fbd6f0d924206d5d012e6da3b72af"
  , "refNftUtxoRef": "39f987a6beb9cc4c45bba149a21c28068f640f3593f15f8157f0b6022b431977#0"
  , "scriptRef": "39f987a6beb9cc4c45bba149a21c28068f640f3593f15f8157f0b6022b431977#1"
  , "nftPolicyRef": "39f987a6beb9cc4c45bba149a21c28068f640f3593f15f8157f0b6022b431977#0"
  }
}
```

**Comment by Micrograx** [comment, marked as answer]
Hi, `refAddr`, `refNftAC` and `refNftUtxoRef` are used as configuration for the actual smart contract address. To configure Kupo you can match everything with `*` or use this address: `addr_test1wqtcz4vq80zxr3dskdcuw7wtfq0vwssd7rrpnnvcvrjhp5sx7leew`.

Keep in mind that the address is derived from the other options so changing them will modify the address. (For example if the validator is ever updated or when it get's deployed to mainnet)

**Reply by sourabhxyz** [reply]
I think that won't be enough, one definitely needs Kupo to keep UTxOs for always fail address where these reference scripts are stored and UTxOs at bot's address to utilize bot's UTxOs.

Anyways, personally I have Kupo synced up with `*` pattern so I prefer just that.

**Comment by SmaugPool** [comment]
I think it should be documented in the README, even if it's `--match '*'` at first. 
I did look for the information without success and ended up in this discussion.

**Reply by sourabhxyz** [reply]
Thanks for feedback @SmaugPool, we'll take that into account and update the documentation.

The idea with kupo is that, it will maintain for those UTxOs which we explicitly ask it for. But since it has really low RAM requirements, `*` pattern doesn't hurt. For SORs, if I am not mistaken, minimal set of UTxOs it should maintain would be those belonging to validator script, bot's address, always fail address (as some reference scripts are stored at always fail validator address) - but again, this is something I would have to verify.

**Comment by sourabhxyz** [comment]
@SmaugPool @mrgrauel Kindly note that running Kupo on `*` match pattern would quickly burn out the SSDs TBW! Please see the updated README on how to run Kupo efficiently, a sample script is available here https://github.com/geniusyield/smart-order-router/tree/main/scripts.
